### PR TITLE
feat(composition): Questionnaire de-referencing

### PR DESCRIPTION
**Pull request by nsenave**
## Questionnaire composition

A questionnaire can contain reference to other questionnaires. 
A de-referencing processing replaces the referenced questionnaire by its content (and updates some elements in the referencing questionnaire).

**For now, if a referenced contain references to other questionnaires, these will not be resolved.**

- Nothing added around the de-refrencing processing, see:
  -  #144 
  - #145 
  - #150 
- Updated `dev-compo-visu` branch with `develop` (⚠️ several conflicts resolved)
- Little refactor

### CODE REVIEW

File: src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    public PoguesJSONToPoguesJSONDerefImpl() {}
+
+    public PoguesJSONToPoguesJSONDerefImpl(QuestionnairesService questionnairesService) {
+        this.questionnairesService = questionnairesService;
+    }
+
+    @Override
+    public void transform(InputStream input, OutputStream output, Map<String, Object> params, String surveyName) throws Exception {
+        if (null == input) {
+            throw new NullPointerException(NULL_INPUT_MESSAGE);
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java:43]: Lombok @NonNull ?

File: src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        if (null == output) {
+            throw new NullPointerException(NULL_OUTPUT_MESSAGE);
+        }
+        String jsonDeref = transform(input, params, surveyName);
+        output.write(jsonDeref.getBytes(StandardCharsets.UTF_8));
+    }
+
+    @Override
+    public String transform(InputStream input, Map<String, Object> params, String surveyName) throws Exception {
+        if (null == input) {
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java:54]: Lombok @NonNull ?

File: src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            throw new NullPointerException(NULL_INPUT_MESSAGE);
+        }
+        return transform(IOUtils.toString(input, StandardCharsets.UTF_8), params, surveyName);
+    }
+
+    @Override
+    public String transform(String input, Map<String, Object> params, String surveyName) throws Exception {
+        if (null == input) {
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java:62]: Lombok @NonNull ?

File: src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        if (!(boolean) params.get("needDeref")) {
+            logger.info("No de-referencing needed");
+            return input;
+        }
+        Questionnaire questionnaire = transformAsQuestionnaire(input);
+        return PoguesSerializer.questionnaireJavaToString(questionnaire);
+    }
+
+    public Questionnaire transformAsQuestionnaire(String input) throws Exception {
+        if (null == input) {
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java:76]: Lombok @NonNull ?

File: src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        return questionnaire;
+    }
+
+    private void deReference(List<String> references, Questionnaire questionnaire) throws Exception {
+        for (String reference : references) {
+            JSONObject referencedJsonQuestionnaire = questionnairesService.getQuestionnaireByID(reference);
+            if (referencedJsonQuestionnaire == null) {
+                logger.warn(
+                        "Null reference behind reference '{}' in questionnaire '{}'.",
+                        reference, questionnaire.getId());
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java:100]: no need for else here

File: src/main/java/fr/insee/pogues/transforms/visualize/composition/FilterComposition.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+            throws IllegalFlowControlException {
+        //
+        String reference = referencedQuestionnaire.getId();
+        //
+        String[] flowControlBounds = getFlowControlBounds(flowControlType);
+        // Replace questionnaire reference by its first/last sequence
+        String beginMember = flowControlBounds[0];
+        String endMember = flowControlBounds[1];
+        if (beginMember.equals(reference)) {
+            beginMember = referencedQuestionnaire.getChild().get(0).getId();
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/composition/FilterComposition.java:35]: get(0) can(t be null ?

File: src/main/java/fr/insee/pogues/transforms/visualize/composition/LoopComposition.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+            throws IllegalIterationException {
+        //
+        String reference = referencedQuestionnaire.getId();
+        //
+        List<String> iterationBounds = getIterationBounds(iterationType);
+        // Replace questionnaire reference by its first/last sequence
+        String beginMember = iterationBounds.get(0);
+        String endMember = iterationBounds.get(1);
+        if (beginMember.equals(reference)) {
+            iterationBounds.set(0, referencedQuestionnaire.getChild().get(0).getId());
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/composition/LoopComposition.java:35]: get(0) can't be null ?

File: src/main/java/fr/insee/pogues/transforms/visualize/composition/QuestionnaireComposition.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            log.info("Iterations' bounds updated in '{}' when de-referencing '{}'", id, reference);
+        }
+
+        // Add iterations (loops)
+        Questionnaire.Iterations refIterations = referencedQuestionnaire.getIterations();
+        if (refIterations != null) {
+            questionnaire.setIterations(new Questionnaire.Iterations());
+            questionnaire.getIterations().getIteration().addAll(refIterations.getIteration());
+            log.info("Iterations from '{}' inserted in '{}'", reference, id);
+        } else {
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/transforms/visualize/composition/QuestionnaireComposition.java:81]: else not needed

File: src/main/java/fr/insee/pogues/utils/PoguesSerializer.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+    /**
+     * Convert the given questionnaire object in a json string.
+     * @param questionnaire Pogues-Model questionnaire.
+     * @return Questionnaire as json string.
+     */
+    public static String questionnaireJavaToString(Questionnaire questionnaire) {
+        try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
+            JAXBContext context = JAXBContext.newInstance(Questionnaire.class);
+            Marshaller marshaller = context.createMarshaller();
```
Comment (davdarras) [inline src/main/java/fr/insee/pogues/utils/PoguesSerializer.java:28]: maybe you can have a class that create the differents (un)marshallers you need

File: src/test/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImplTest.java [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+    }
+
+    /**
+     * Tested questionnaire: 'lct78jr8'
+     * This questionnaire contains references to the following questionnaires:
+     * - 'l4i3m6qa' (with a filter condition in the root questionnaire)
+     * - 'l6dnlrka' (simple questionnaire)
+     * - 'lct8pcsy' (including loops) */
+    @Test
+    void dereference_filterAndLoop() throws Exception {
```
Comment (davdarras) [inline src/test/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImplTest.java:100]: seems to have multiple different tests here. Could be divide into several methods for each test type

File: src/test/java/fr/insee/pogues/utils/PoguesDeserializerTest.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class PoguesDeserializerTest {
+
+    @Test
+    void deserialize_invalidString() throws ParseException {
+        //
```
Comment (davdarras) [inline src/test/java/fr/insee/pogues/utils/PoguesDeserializerTest.java:17]: really minor thing, could have been parameterized

File: src/test/java/fr/insee/pogues/utils/PoguesModelUtilsTest.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import fr.insee.pogues.model.FlowControlType;
+import fr.insee.pogues.model.IterationType;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+class PoguesModelUtilsTest {
+
+    @Test
+    void getFlowControlBounds_nullIfTrue() {
```
Comment (davdarras) [inline src/test/java/fr/insee/pogues/utils/PoguesModelUtilsTest.java:15]: method name does not reflect what it does

File: src/test/resources/transforms/PoguesJSONToPoguesJSONDeref/filter_and_loop/l4i3m6qa.json [hunk 13]
```diff
@@ -0,0 +1,355 @@
...
+  ],
+  "childQuestionnaireRef": [],
+  "Name": "CATALOGUE",
+  "Variables": {
+    "Variable": [
+      {
+        "Formula": "42",
+        "Label": "CAT_VAR_COMP_1",
+        "id": "l50tn9cp",
+        "type": "CalculatedVariableType",
```
Comment (davdarras) [inline src/test/resources/transforms/PoguesJSONToPoguesJSONDeref/filter_and_loop/l4i3m6qa.json:33]: that reminds me of something ... 🤣

**Review by davdarras — COMMENTED** [review]
Good for me !

File: src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java (hunk 5 above)
Comment (nsenave) [reply src/main/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImpl.java:100]: In the [previous implementation](https://github.com/InseeFr/Pogues-Back-Office/commit/8c51caa2d0cb96d9dfaf9db7eeb22df4cc4ee89c) such case would actually have thrown a NullPointerException, I'll remove the else and replace the log.warn with an exception.

File: src/test/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImplTest.java (hunk 10 above)
Comment (nsenave) [reply src/test/java/fr/insee/pogues/transforms/visualize/PoguesJSONToPoguesJSONDerefImplTest.java:100]: Tests in this class are more like integration tests for the de-referencing feature (given some questionnaire files, test as much things as possible with their content). Unit tests are on the `QuestionnaireComposition` class that holds the logic.

File: src/test/java/fr/insee/pogues/utils/PoguesModelUtilsTest.java (hunk 12 above)
Comment (nsenave) [reply src/test/java/fr/insee/pogues/utils/PoguesModelUtilsTest.java:15]: The property name is actually confusing here, I added a comment

**Pull request merged**